### ISSUE TITLE
Iterator bug: undefined constant U

### ISSUE THREAD

**Issue by kostya** [crystal-lang/crystal#2411]
https://play.crystal-lang.org/#/r/w21

```crystal
record Bla, bla

class I
  include Iterator(Bla)
  
  def next
    Bla.new(rand)
  end
end

p I.new.select { |x| x.bla < 0.5 }.map(&.bla).first
```

**Comment by asterite** [comment]
Thanks for reporting.

This bug will be fixed once we require types for instance variables. Then you'll be forced to specify the type of `Bla`'s `@bla` and the compiler will know the type of `x.bla`. The main issue is that the compiler can't know it because you never instantiate `Bar`. Well, you do, but in order to reach that point it has to solve `select` first, so it's kind of recursive and impossible.

Workaround is, do something like:

```crystal
record Bla, bla : Float64
```

(note that this workaround will be the only solution after some versions)

**Comment by kostya** [comment]
it was reduced example:
original example harder:

https://github.com/kostya/myhtml
```crystal
require "./src/myhtml"

parser = Myhtml::Parser.new
parser.parse("<html>bla</html>")
p parser.tags(:_text).select { |n| n.parents.all? &.visible? }.map { |n| n.tag_text.strip }.first
```

```
Error in ./1.cr:5: instantiating 'Iterator::Select(Myhtml::EachTagIterator, Myhtml::Node, Bool)#map()'

p parser.tags(:_text).select { |n| n.parents.all? &.visible? }.map { |n| n.tag_text.strip }.first
                                                               ^~~

in /usr/local/Cellar/crystal-lang/0.15.0/src/iterator.cr:542: undefined constant U

  def map(&func : T -> U)
                       ^
```

not understand how to workaround this

**Comment by asterite** [comment]
The workaround is:

```crystal
module Myhtml
  struct Node
    # Add this line
    @node : Lib::MyhtmlTreeNodeT*
```

The compiler doesn't know the type of `@node` so it can't type `tag_text_slice`, so it can't type `tag_text`.

The good thing is that Node has this constructor:

```crystal
    def initialize(@tree : Tree, @node : Lib::MyhtmlTreeNodeT*)
```

which means that [in a next version](https://github.com/crystal-lang/crystal/issues/2390) the compiler won't complain anymore without you having to change anything :-)

**Comment by kostya** [comment]
thanks

**Comment by kostya** [comment]
this example now looks so nice: https://github.com/kostya/myhtml/blob/master/examples/texts.cr#L33
btw, crystal wrong format this place!

**Comment by asterite** [comment]
This is fixed now because of #2443 (the original snippet doesn't compile, and it works fine if Bla#@bla is given a Float64 type).
